File: WolfDen/Game/Other/RepairCoordinator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using WolfDen.Game.EventArguments;

namespace WolfDen.Game.Other
{
    public class RepairCoordinator
    {
        public string Class { get { return GetType().Name; } }
        public AdvanceWarsPlayer Player { get; set; }
        public int TurnRepairCosts { get; set; }
        public event Action CalcRepairCostsEvent;
        public event Action ExecuteRepairsEvent;
        public IncomeCoordinator IncomeCoordinator { get; set; }


        public RepairCoordinator(AdvanceWarsPlayer player)
        {
            Player = player;
        }

        public void Initialize(IncomeCoordinator incomeCoordinator)
        {
            IncomeCoordinator = incomeCoordinator;
            incomeCoordinator.PostIncomeEvent += CoordinateRepairs;
        }

        private void CoordinateRepairs(object sender, ListOfPlayersEventArgs args)
        {
            if (!args.Players.Contains(Player))
                return;

            TurnRepairCosts = 0;
            RaiseCalcRepairCostsEvent();
            if (TurnRepairCosts <= Player.Gold)
                RaiseExecuteRepairsEvent();
            TurnRepairCosts = 0;
        }

        private void RaiseCalcRepairCostsEvent() => CalcRepairCostsEvent?.Invoke();
        private void RaiseExecuteRepairsEvent() => ExecuteRepairsEvent?.Invoke();
    }
}

File: WolfDen/wwwroot/ts/Other/RepairCoordinator.ts
namespace TurnBasedBoardGame {
    export class RepairCoordinator {
        Player: AdvanceWarsPlayer;
        TurnRepairCosts: number = 0;
        CalcRepairCostsEvent: any = {};
        ExecuteRepairsEvent: any = {};
        IncomeCoordinator: IncomeCoordinator;
        Id: string = guid();

        Initialize() {
            this.IncomeCoordinator.PostIncomeEvent[this.Id] = (players) => { this.CoordinateRepairs(players); };
        }

        CoordinateRepairs(activePlayers: AdvanceWarsPlayer[]) {
            if (activePlayers.indexOf(this.Player) == -1)
                return;

            this.TurnRepairCosts = 0;
            this.RaiseCalcRepairCostsEvent();
            if (this.TurnRepairCosts <= this.Player.Gold)
                this.RaiseExecuteRepairsEvent();
            this.TurnRepairCosts = 0;
        }

        RaiseCalcRepairCostsEvent() {
            for (let key in this.CalcRepairCostsEvent) {
                if (typeof this.CalcRepairCostsEvent[key] === "function")
                    this.CalcRepairCostsEvent[key]();
            }
        }

        RaiseExecuteRepairsEvent() {
            for (let key in this.ExecuteRepairsEvent) {
                if (typeof this.ExecuteRepairsEvent[key] === "function")
                    this.ExecuteRepairsEvent[key]();
            }
        }

    }
}